File: docs/js/mechanics/buff.js
/**
 * Buff / Debuff system.
 *
 * Each buff has:
 *   id        — unique identifier
 *   name      — display name
 *   icon      — emoji
 *   stacks    — current stack count (≥1)
 *   maxStacks — max allowed stacks (0 = unlimited)
 *   duration  — remaining rounds (0 = permanent until removed)
 *   type      — 'buff' | 'debuff'
 *   description(stacks) — returns text for current stack count
 *
 * Registered buff definitions live in BUFF_DEFS.
 */

/** Buff / debuff instance attached to a combatant. */
export class Buff {
    constructor(def, stacks = 1, duration = 0) {
        this.id = def.id;
        this.def = def;
        this.stacks = Math.min(stacks, def.maxStacks || Infinity);
        this.duration = duration; // 0 = permanent
    }

    get name() { return this.def.name; }
    get icon() { return this.def.icon; }
    get type() { return this.def.type; }

    /** Add stacks (clamped to maxStacks). */
    addStacks(n) {
        const max = this.def.maxStacks || Infinity;
        this.stacks = Math.min(this.stacks + n, max);
    }

    /** Tick duration. Returns true if expired. */
    tick() {
        if (this.duration <= 0) return false; // permanent
        this.duration -= 1;
        return this.duration <= 0;
    }

    /** Get description for current stacks. */
    describe() {
        return this.def.description(this.stacks);
    }
}

/* ═══════════════════════════════════════════
 *  Buff Definitions Registry
 * ═══════════════════════════════════════════ */

export const BUFF_DEFS = {};

function registerBuff(def) {
    BUFF_DEFS[def.id] = def;
}

/* ─── 滋润 (Nourish) ─── */
registerBuff({
    id: 'nourish',
    name: '滋润',
    icon: '💧',
    type: 'buff',
    maxStacks: 20,
    description: (stacks) => `回复效果 +${stacks * 5}%`,
    /** Returns the heal multiplier bonus for current stacks (e.g. 2 stacks → 0.10). */
    healBonus: (stacks) => stacks * 0.05,
});

/* ─── 脆弱 (Fragile) ─── */
registerBuff({
    id: 'fragile',
    name: '脆弱',
    icon: '🔻',
    type: 'debuff',
    maxStacks: 20,
    description: (stacks) => `受到伤害 +${stacks * 5}%`,
    /** Returns the incoming damage multiplier bonus (e.g. 2 stacks → 0.10). */
    damageIncrease: (stacks) => stacks * 0.05,
});

/* ═══════════════════════════════════════════
 *  Buff Manager — manages buffs on a combatant
 * ═══════════════════════════════════════════ */

export class BuffManager {
    constructor() {
        /** @type {Buff[]} */
        this.buffs = [];
    }

    /** Apply or stack a buff by id. */
    apply(id, stacks = 1, duration = 0) {
        const def = BUFF_DEFS[id];
        if (!def) return null;

        const existing = this.buffs.find(b => b.id === id);
        if (existing) {
            existing.addStacks(stacks);
            // Refresh duration if new duration is longer
            if (duration > 0 && duration > existing.duration) {
                existing.duration = duration;
            }
            return existing;
        }

        const buff = new Buff(def, stacks, duration);
        this.buffs.push(buff);
        return buff;
    }

    /** Remove a buff by id. */
    remove(id) {
        this.buffs = this.buffs.filter(b => b.id !== id);
    }

    /** Get a buff by id (or null). */
    get(id) {
        return this.buffs.find(b => b.id === id) || null;
    }

    /** Get current stacks of a buff (0 if absent). */
    stacks(id) {
        return this.get(id)?.stacks || 0;
    }

    /** Tick all buff durations, removing expired ones. Returns list of expired buff ids. */
    tick() {
        const expired = [];
        this.buffs = this.buffs.filter(b => {
            if (b.tick()) {
                expired.push(b.id);
                return false;
            }
            return true;
        });
        return expired;
    }

    /** Clear all buffs. */
    clear() {
        this.buffs = [];
    }

    /** Whether any buffs exist. */
    get hasBuff() {
        return this.buffs.length > 0;
    }
}

File: src/mechanics/buff.rs
/// Buff / Debuff system.
/// Known buff identifiers.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum BuffId {
    /// 滋润 — each stack +5% heal effectiveness.
    Nourish,
    /// 脆弱 — each stack +5% incoming damage.
    Fragile,
}

impl BuffId {
    pub fn name(&self) -> &'static str {
        match self {
            BuffId::Nourish => "滋润",
            BuffId::Fragile => "脆弱",
        }
    }

    pub fn icon(&self) -> &'static str {
        match self {
            BuffId::Nourish => "💧",
            BuffId::Fragile => "🔻",
        }
    }

    pub fn max_stacks(&self) -> u32 {
        match self {
            BuffId::Nourish => 20,
            BuffId::Fragile => 20,
        }
    }
}

/// A buff/debuff instance on a combatant.
#[derive(Debug, Clone)]
pub struct Buff {
    pub id: BuffId,
    pub stacks: u32,
    pub duration: u32, // 0 = permanent
}

impl Buff {
    pub fn new(id: BuffId, stacks: u32, duration: u32) -> Self {
        let max = id.max_stacks();
        Self {
            id,
            stacks: stacks.min(max),
            duration,
        }
    }

    pub fn add_stacks(&mut self, n: u32) {
        let max = self.id.max_stacks();
        self.stacks = (self.stacks + n).min(max);
    }

    /// Tick duration. Returns true if expired.
    pub fn tick(&mut self) -> bool {
        if self.duration == 0 {
            return false;
        }
        self.duration = self.duration.saturating_sub(1);
        self.duration == 0
    }
}

/// Manages buffs on a combatant.
#[derive(Debug, Clone, Default)]
pub struct BuffManager {
    pub buffs: Vec<Buff>,
}

impl BuffManager {
    pub fn new() -> Self {
        Self { buffs: Vec::new() }
    }

    pub fn apply(&mut self, id: BuffId, stacks: u32, duration: u32) {
        if let Some(existing) = self.buffs.iter_mut().find(|b| b.id == id) {
            existing.add_stacks(stacks);
            if duration > 0 && duration > existing.duration {
                existing.duration = duration;
            }
        } else {
            self.buffs.push(Buff::new(id, stacks, duration));
        }
    }

    pub fn remove(&mut self, id: BuffId) {
        self.buffs.retain(|b| b.id != id);
    }

    pub fn stacks(&self, id: BuffId) -> u32 {
        self.buffs
            .iter()
            .find(|b| b.id == id)
            .map_or(0, |b| b.stacks)
    }

    /// Tick all buffs and remove expired ones. Returns expired ids.
    pub fn tick(&mut self) -> Vec<BuffId> {
        let mut expired = Vec::new();
        self.buffs.retain_mut(|b| {
            if b.tick() {
                expired.push(b.id);
                false
            } else {
                true
            }
        });
        expired
    }

    pub fn clear(&mut self) {
        self.buffs.clear();
    }

    /// Calculate heal multiplier from nourish buff (1.0 + stacks * 0.05).
    pub fn nourish_heal_multiplier(&self) -> f64 {
        let stacks = self.stacks(BuffId::Nourish);
        1.0 + stacks as f64 * 0.05
    }

    /// Calculate incoming damage multiplier from fragile debuff (1.0 + stacks * 0.05).
    pub fn fragile_damage_multiplier(&self) -> f64 {
        let stacks = self.stacks(BuffId::Fragile);
        1.0 + stacks as f64 * 0.05
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn nourish_heal_bonus() {
        let mut mgr = BuffManager::new();
        assert!((mgr.nourish_heal_multiplier() - 1.0).abs() < f64::EPSILON);

        mgr.apply(BuffId::Nourish, 2, 0);
        assert!((mgr.nourish_heal_multiplier() - 1.10).abs() < f64::EPSILON);

        mgr.apply(BuffId::Nourish, 3, 0);
        assert_eq!(mgr.stacks(BuffId::Nourish), 5);
        assert!((mgr.nourish_heal_multiplier() - 1.25).abs() < f64::EPSILON);
    }

    #[test]
    fn max_stacks_capped() {
        let mut mgr = BuffManager::new();
        mgr.apply(BuffId::Nourish, 100, 0);
        assert_eq!(mgr.stacks(BuffId::Nourish), 20);
    }

    #[test]
    fn fragile_damage_bonus() {
        let mut mgr = BuffManager::new();
        assert!((mgr.fragile_damage_multiplier() - 1.0).abs() < f64::EPSILON);

        mgr.apply(BuffId::Fragile, 3, 0);
        assert!((mgr.fragile_damage_multiplier() - 1.15).abs() < f64::EPSILON);

        mgr.apply(BuffId::Fragile, 2, 0);
        assert_eq!(mgr.stacks(BuffId::Fragile), 5);
        assert!((mgr.fragile_damage_multiplier() - 1.25).abs() < f64::EPSILON);
    }

    #[test]
    fn fragile_max_stacks_capped() {
        let mut mgr = BuffManager::new();
        mgr.apply(BuffId::Fragile, 100, 0);
        assert_eq!(mgr.stacks(BuffId::Fragile), 20);
    }

    #[test]
    fn duration_tick_expires() {
        let mut mgr = BuffManager::new();
        mgr.apply(BuffId::Nourish, 3, 2);
        assert_eq!(mgr.stacks(BuffId::Nourish), 3);

        let expired = mgr.tick();
        assert!(expired.is_empty());
        assert_eq!(mgr.stacks(BuffId::Nourish), 3);

        let expired = mgr.tick();
        assert_eq!(expired, vec![BuffId::Nourish]);
        assert_eq!(mgr.stacks(BuffId::Nourish), 0);
    }
}
